### PR TITLE
Add mindstream recipe

### PR DESCRIPTION
**Pull request by countvajhula**
### Brief summary of what the package does

Mindstream allows you to enter freewriting sessions for writing or coding. It starts off like a simple scratch buffer that you don't even need to name, but the buffer is Git-versioned and committed every time it's saved, giving rise to robust sessions that remember everything you did, so that you never need to worry about losing anything even though it was so easy to get started (e.g. you didn't have to think of a filename, location on disk, or anything). When you're done, you can save these sessions to a permanent location and resume them at any time, or just close them and forget about them if their purpose has been served.

Example usage:

- for quick prototyping in any programming language
- learning a new programming concept
- solving coding problems with a quick feedback loop
- writing a blog post or even a book
- drafting emails
- really any writing task (like this description!)

### Direct link to the package repository

https://github.com/countvajhula/mindstream

### Your association with the package

Maintainer

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

### CODE REVIEW

**Comment by walseb** [comment]
Cool package! 

It's great to see a package improving the backup functionality of Emacs beyond backup files and saved undo state.

Here are some thoughts unrelated to whether this should be on Melpa or not:

For the backup part, maybe you'd want to use the API provided by git-backup? https://melpa.org/#/git-backup
But that might be a lot of work.

I see that you are recommending git-timemachine to traverse the various iterations committed to the git repo. Last time I looked into it, it didn't show you a diff of the changes made compared to your current buffer, but just what the selected git commit looks like, which I found very confusing as it's hard to see sometimes what's changed.

I suspect also that restoring a commit destroys any overlays you have created, moves point, etc, which is very annoying.

I have been trying to fix this in [my interface](https://melpa.org/#/git-backup-ivy) to [git-backup](https://melpa.org/#/git-backup), which I think does the backup part of this package, but globally. I have the live diff working well even for large changes (with some bugs), but haven't gotten the restoring part working. There's this package that I can't remember the name of that can restore markers and such, but it's made for formatters only. I have been working a bit on adapting it, but haven't gotten it to work.

In conclusion, would be great if we had a better git-timemachine that could do previews and restore git commits without moving around all the Emacs state in the buffer.

**Comment by countvajhula** [comment]
Hi! From a quick look, git-backup isn't a good fit for the backend (though it's cool). Mindstream is not really about backups but about sessions. It's like:

- Vim's undotree, except based on Git (with all the benefits that brings), and supporting more than one file
- Persistent scratch buffers like persistent-scratch, except versioned and with an emphasis on scaling with your work rather than only on starting and for scratch notes
- Like DreamPie, GroovyConsole, DrRacket -- and many other IDEs that offer an interactive scratch buffer for an integrated development feedback loop. Except that the scratch buffer becomes the project and you don't need to create new files for transitioning to "real work"

Emacs's tooling around Git (like Git-timemachine and Magit) is very handy and clever, and there's surely going to be a lot more over time (like your diff viewer!). That's a big reason I chose to implement the sessions using Git, so that any such common Git tools written in the community would be usable together with Mindstream sessions out of the box!

**Comment by riscy** [comment]
Thanks both - conversations like these make some PRs especially fun for me to look at.

At a glance, one thing I would change is to have `mindstream-backend-initialize` use `shell-quote-argument` on `base-path` just in case `base-path` could contain any shell control characters, which the user could introduce if they set e.g. `mindstream-path` to "riscy's stuff" or something similar.

Alternatively instead of `shell-command-to-string` you might want to use `call-process`, per the docs --
```
In Elisp, you will often be better served by calling ‘call-process’ or
‘start-process’ directly, since they offer more control and do not
impose the use of a shell (with its need to quote arguments).
```

**Comment by countvajhula** [comment]
Have you considered broadening the scope of your tool outside a git-backups setting (e.g. as an addon to git-timemachine) so that it's usable with Git in general? That could make it even more useful.

Thanks for the feedback as always @riscy . I'm tied up for a couple of days, but I should be able to follow up here later this week.

**Comment by countvajhula** [comment]
I've made the change you suggested @riscy . Now that I'm looking at it again, it would be nice to do this via standard Elisp bindings for Git here instead of using the CLI. Magit APIs are too user focused / interactive for use here, but [libegit2](https://github.com/magit/libegit2) seemed promising. It doesn't look like that package is quite ready though.

**Comment by riscy** [comment]
Thank you!  Here's the latest --

`mindstream-backend.el` with byte-compile using Emacs 29.3:
```
mindstream-backend.el:63:9: Warning: reference to free variable `mindstream-add-everything'
```

Package and license:
- Other (via API)
  - Try to use a standard license file format
    See: https://github.com/licensee/licensee
    e.g. the fulltext is here https://spdx.org/licenses/Unlicense.html

**Comment by countvajhula** [comment]
Ah whoops, I've addressed the warning you found.

I also realized that I needed to make a change to how multiple files would be supported that would have been backwards incompatible, so I've gone ahead and preemptively made this change to avoid that. To summarize the change, formerly Mindstream was oriented around a version-controlled _buffer_. But with "sessions" being generalized to full Git repos with any files contained there, it was necessary to make templates into _folders_ rather than files, and that involved a lot of rewiring. It was also necessary to eliminate the buffer-local `mindstream-session-mode` and instead have everything happen via the global `mindstream-mode` along with a registry of "active sessions" that are simply Git repo paths.

It should be good to go now, and ready for any further feedback!

re: the license format, if you don't mind, I prefer to retain the COPYING file that simply indicates that the package is public domain and there's no copyright. The reason is that blindly placing off-the-shelf licenses on software (as we've been taught to do) is a misguided habit that distracts from the real problems to be solved here -- which is to say, _attribution_ of work, and consequent financial _empowerment_ of the creators of value. Licenses based in outdated notions of ownership of software, in the form of copyright, are not going to get us there, but the simple use of public domain, together with fair attribution processes governed by sound economic principles -- and if we support one another -- will.

On this subject, I humbly urge you to watch [my talk at EmacsConf 2022](https://emacsconf.org/2022/talks/maint/), if you have a moment. It is an important one on the subject of open source sustainability and the path to making economic processes function in an open and transparent way, retaining everything that's good about open source and gaining everything that's good about the traditional economic processes that actually work today. 🙏

**Comment by countvajhula** [comment]
Hey @riscy , not to rush you, but JFYI I'm going to be demoing Mindstream at a meetup this coming Saturday. If you feel it's ready, it'd be convenient if it could be merged before then so that attendees could try it out -- but no pressure at all (and it would be valuable -- as always! -- to hear further feedback).

Re: the remaining byte compile warning from melpazoid that I see:

```
mindstream-session.el:76:[29](https://github.com/countvajhula/mindstream/actions/runs/8919607680/job/24496084851#step:5:30): Warning: `lax-plist-get' is an obsolete function (as of 29.1); use `plist-get' instead.
```

As it happens, I'm using Emacs 27.1 at the moment and it doesn't have the new `predicate` argument to `plist-get` that allows the use of `equal` instead of `eq`. Since Mindstream only depends on Emacs 25.1 at this point, it seems premature to transition to `plist-get`, but what do you think? I could also aim to make the change after I personally upgrade my Emacs.

**Comment by riscy** [comment]
Darnit, I sincerely apologize for not seeing that in time.  I'm pretty much only looking at MELPA in batch on my weekends (and often just on Sundays) which includes notifications and emails.  Still, I regret I wasn't able to help.

Regarding the license, my impression was you were using the Unlicense because you've used exactly its phrasing, i.e. `This is free and unencumbered software released into the public domain` and indeed the Unlicense is a public domain dedication.  I think if you want to strictly distance yourself from an off-the-shelf license like the Unlicense, as you write above, then you should not re-use their unique phrasing - just to prevent confusion.

You have obviously thought deeply about licenses, but I should also note that the notion of "public domain" doesn't necessarily exist in all legal jurisdictions where this software might be used. Nevertheless public domain *is* GPL compatible.

Altogether I think this is only a trivial issue, and the code looks good now.  Merging!

**Comment by countvajhula** [comment]
No worries at all, it's much more important for you to contribute at a healthy cadence that works for you. Thank you for your dedicated service to the community!

> Regarding the license, my impression was you were using the Unlicense because you've used exactly its phrasing, i.e. This is free and unencumbered software released into the public domain and indeed the Unlicense is a public domain dedication. I think if you want to strictly distance yourself from an off-the-shelf license like the Unlicense, as you write above, then you should not re-use their unique phrasing - just to prevent confusion.

Ah, I understand, and I apologize for the confusion there! Honestly, it's a delicate balance to strike between satisfying expectations of being legally clear cut (hence similarity of verbiage to existing licenses, even [with an assent form](https://github.com/countvajhula/mindstream/blob/main/pull_request_template.md) for every contributor in PRs) while also not being a license at all but a declaration that the very presumption of licenses in such instances -- that you own something you are in a position to lend to others under certain conditions -- is wrong. To elaborate on this a bit, it's a big leap and not easy to understand why this makes sense, but it ultimately comes down to what ownership even means and it necessarily touches upon some profound philosophical currents like the finiteness of life, the concept of society, and the ultimate basis for notions of possession. In ABE, the conclusion is that ownership only makes sense as the _institution_ of property, and its basis is _not_ in something inherent like "I made this, therefore it is mine," but rather in the concept of _agreement_ by society. It is because a general notion of property is beneficial _to all_ that we institute it _collectively_ for any individual. This basis shows us how property is inapplicable to software, or rather, that it is a very heavy-handed instrument to apply that ends up _severely limiting_ value creation along with _severely skewing_ who gets empowered by dint of what value is generated (consider those who are working on valuable AI tech behind closed doors), thus, not serving the common interest at all.

The GPL's answer is an economic boundary condition where it solves the problem by eliminating all of the real complexity attending the problem in the first place -- richness that matters is simply unmodeled. It does not account for attribution at all, so that, in the mathematical limit, voices in a GPL society are not empowered by the value they contribute to others. But what is presumed to be gained is that no one can constrain the freedom of others. It is an admirable and courageous effort that set us on the right path, but it is ultimately not a better solution than the one (proprietary capitalism) it was proposed in opposition to, since leaving value unmodeled is tantamount to a presumption of "equality" which, although it sounds good, is not a concept that makes sense either economically or philosophically and does not ultimately get us what we really want. Really, we seek other goals that this illusory term masks -- goals like fairness, caring for one another, innovation and access to good things, harmony, community, love, kindness. That's what we are really after, and there is only _one_ way to achieve this in a sustainable way at arbitrary scale, and that is to develop a _sound economic theory_ because this governs the dynamics of what we are.

The real question we seek to answer is, can we maximize value generation (i.e. "public domain" -- no control _at all_, not even "to ensure no one controls") while also fairly empowering the sources of value (attribution), without using a costly, demeaning institution like property (which, in addition to being a huge drag on open source in the form of "license soup," is also the basis of proprietary software -- so ironically, placing even "free" licenses on open source software admits the legitimacy of proprietary software!)? ABE is engaged in developing the affirmative answer to this question, by means of a wonderful "mirror" at the heart of the theory that allows us to act in our own interest by acting in the collective interest (indeed, in ABE it is not _required_ for you to use public domain, but it is _in your best interests_ to do so!).

That is certainly long winded, but I wanted to share you with the challenges of striking a balance between satisfying people's expectations today while also unraveling them to get us to a better place. I'll gladly incorporate any feedback that might help us do this more effectively, and thanks again!

**Comment by riscy** [comment]
That *was* a lot, but I promise I read it - and thank you for taking the time.  I agree especially with the point about license soup and open licenses "legitimizing" (or at least working in a compatible way with) the system.  But - to play devil's advocate - I also think that these are the only tools and concepts that our legal system uses when determining the penalties that can very much affect our finite lives.  Maybe I am just skewed to short-term thinking though.

Incidentally this post does not necessarily represent my employer's positions, strategies, or opinions.

**Comment by countvajhula** [comment]
Indeed. That's why it's important to raise awareness. It is not through legal debate founded on improper assumptions but through adoption and experiencing better alternatives that we'll have a sea change.

**Comment by countvajhula** [comment]
I felt inspired: [Copyright v. Attribution Based Economics](https://drym.org/copyright-v-attribution-based-economics/)
